lib/token-store.js

```diff
@@ -8,196 +8,1110 @@
+     * @param {number} beforeCount - The number of tokens this cursor iterates before start.
+     * @param {number} afterCount - The number of tokens this cursor iterates after end.
      */
-    api.getTokenBefore = function(node, skip) {
-        return tokens[starts[node.range[0]] - (skip || 0) - 1];
-    };
+    constructor(tokens, comments, imap, start, end, beforeCount, afterCount) {
+        super(tokens, comments, imap, start, end);
+        this.beforeCount = beforeCount | 0;
+        this.afterCount = afterCount | 0;
+    }
+
+    /** @inheritdoc */
+    get firstIndex() {
+        return Math.max(0, super.firstIndex - this.beforeCount);
+    }
+
+    /** @inheritdoc */
+    get lastIndex() {
+        return Math.min(this.tokens.length - 1, super.lastIndex + this.afterCount);
+    }
+}
+
+/**
+ * The abstract class about cursors which manipulate another cursor.
+ * @private
+ */
+class DecorativeCursor extends Cursor {
 
     /**
-     * Gets a number of tokens that follow a given node or token in the token
-     * stream.
-     * @param {(ASTNode|Token)} node The AST node or token.
-     * @param {int} [afterCount=0] The number of tokens after the node or token
-     *     to retrieve.
-     * @returns {Token[]} Array of objects representing tokens.
+     * Initializes this cursor.
+     * @param {Cursor} cursor - The cursor to be decorated.
      */
-    api.getTokensAfter = function(node, afterCount) {
-        const start = lastTokenIndex(node) + 1;
+    constructor(cursor) {
+        super();
+        this.cursor = cursor;
+    }
 
-        return get(start, start + (afterCount || 0));
-    };
+    /** @inheritdoc */
+    moveNext() {
+        const retv = this.cursor.moveNext();
+
+        this.current = this.cursor.current;
+
+        return retv;
+    }
+
+    /** @inheritdoc */
+    movePrev() {
+        const retv = this.cursor.movePrev();
+
+        this.current = this.cursor.current;
+
+        return retv;
+    }
+}
+
+/**
+ * The decorative cursor which ignores specified tokens.
+ * @private
+ */
+class FilterCursor extends DecorativeCursor {
 
     /**
-     * Gets the token that follows a given node or token in the token stream.
-     * @param {(ASTNode|Token)} node The AST node or token.
-     * @param {int} [skip=0] A number of tokens to skip after the given node or
-     *     token.
-     * @returns {Token} An object representing the token.
+     * Initializes this cursor.
+     * @param {Cursor} cursor - The cursor to be decorated.
+     * @param {Function} predicate - The predicate function to decide tokens this cursor iterates.
      */
-    api.getTokenAfter = function(node, skip) {
-        return tokens[lastTokenIndex(node) + (skip || 0) + 1];
-    };
+    constructor(cursor, predicate) {
+        super(cursor);
+        this.predicate = predicate;
+    }
+
+    /** @inheritdoc */
+    moveNext() {
+        while (super.moveNext()) {
+            if (this.predicate(this.current)) {
```

Comment:
It's only a minor issue, but `predicate` will get called with a `this` value of the current instance of `FilterCursor`, which would be observable to external code. This could make it difficult to change the implementation in the future without making breaking changes to rules.

Maybe we should call the predicate as a pure function:

```js
const predicate = this.predicate;

while (super.moveNext()) {
    if (predicate(this.current)) {
        // ...
```